Provide a .spec file to build RPM packages

figless opened the pull request:
This MR provides basic support for packaging `autopatch` as an RPM via `setup.py`. 
A `Makefile` is also provided to facilitate building `srpm` and `rpm` targets.

The RPM is defined as a metapackage for `autopatch`, with sub packages which have additional functionality split out. For the most basic use-case, the `autopatch-core` package can be used.

For full functionality the packages `autopatch-ansible`, `autopatch-git`, `autopatch-slack` and `autopatch-web` are provided.

In addition a simple `autopatch_standalone.py` script has been provided which can be used without any git or web components (provided via `autopatch-core`) and deployed on the system in `%{bindir}/autopatch`

Notes:
- Currently it's impossible to install `autopatch-git` due to missing dependencies (https://github.com/AlmaLinux/immudb-wrapper and the upstream `immudb-py` client) should be packaged
- `autopatch-slack` requires a slack python module that exists in Fedora, but not currently in EPEL9

figless committed:
Update source to try importing packaged path, with failback to local

figless committed:
Add Makefile

figless committed:
Add setup.py

figless committed:
Add autopatch.spec

figless committed:
Add autopatch_standalone.py

figless committed:
Add some documentation about the spec file to README.md

eabdullin1 commented:
I believe that the **web** subpackage should have a requirement on **slack** subpackage, as it uses **tools_slack**, and also on **git** subpackage, as `src/debranding.py` uses the **git** module.
I’m not sure that it should provide the **ansible** subpackage, do we have any reason to include it?

Here is the testing build:
[https://build.almalinux.org/build/31562](https://build.almalinux.org/build/31562)

figless committed:
Add autopatch-{git,slack} dependencies to autopatch-web

figless committed:
Remove the ansible subpackage

figless commented:
> I believe that the **web** subpackage should have a requirement on **slack** subpackage, as it uses **tools_slack**, and also on **git** subpackage, as `src/debranding.py` uses the **git** module. I’m not sure that it should provide the **ansible** subpackage, do we have any reason to include it?
> 
> Here is the testing build: https://build.almalinux.org/build/31562

I've added those additional `Requires`. Packaging `ansible` is not mandatory, i've removed the `ansible` subpackage

figless committed:
Ensure autopatch-{git,slack} dependencies are defined correctly

eabdullin1 commented:
Great. I’ll merge and build it after the **slackclient** and **immudb** packages.
This will probably happen after the 9.6 release.
But I will try to do that ASAP

eabdullin1 commented:
@figless Hey, I finally had some time. I’ve created test builds. Could you please check if they cover all your needs?
https://build.almalinux.org/build/40117
https://build.almalinux.org/build/40118
https://build.almalinux.org/build/40187
https://build.almalinux.org/build/40210

figless commented:
> @figless Hey, I finally had some time. I’ve created test builds. Could you please check if they cover all your needs? https://build.almalinux.org/build/40117 https://build.almalinux.org/build/40118 https://build.almalinux.org/build/40187 https://build.almalinux.org/build/40210

Looks good to me. The `autopatch-core` package is enough for our use-case anyway.
We can now probably remove some of the comments in the spec file related to package requirements. There is still the requirement on `python3-slackclient` which would need to be built in EPEL8/9/10 (it does exist in Fedora). How do you want to deal with that?

eabdullin1 commented on autopatch.spec:
```diff
@@ -0,0 +1,123 @@
+Requires:       python3-requests
+Requires:       git
+# python3-immudb-wrapper needs to be packaged (as well as the upstream python3-immudb)
+Requires:       python3-immudb-wrapper
```
Could you please rename it to python3-immudb_wrapper

figless replied:
renamed :)

eabdullin1 commented:
> > @figless Hey, I finally had some time. I’ve created test builds. Could you please check if they cover all your needs? https://build.almalinux.org/build/40117 https://build.almalinux.org/build/40118 https://build.almalinux.org/build/40187 https://build.almalinux.org/build/40210
> 
> Looks good to me. The `autopatch-core` package is enough for our use-case anyway. We can now probably remove some of the comments in the spec file related to package requirements. There is still the requirement on `python3-slackclient` which would need to be built in EPEL8/9/10 (it does exist in Fedora). How do you want to deal with that?

Great.
I already built python3-`slackclient`. I just forgot to add build
here it is https://build.almalinux.org/build/31604

eabdullin1 commented:
Ok, one more thing. If you’re fine with just the core packages, I propose dropping all other subpackages (web, git, slack) and keep only core

figless committed:
Revert git,web,slack subpackages and provide autopatch,autopatch-core

figless commented:
> Ok, one more thing. If you’re fine with just the core packages, I propose dropping all other subpackages (web, git, slack) and keep only core

I have refactored the spec file.

Now we have an `autopatch` package which depends on `autopatch-core` and installs the dependencies and `%file` defines for the git,web,slack components.

`autopatch-core` is the core package which can be used without requiring installation of git,slack,web codebase and dependencies.

figless commented:
On reflection, perhaps `autopatch-minimal` is a better name than `autopatch-core`. What do you think ?

eabdullin1 commented:
I prefer to go with autopatch-core. I will rebuild it and share new build with you.
If it works fine, we can release only core package to the Synergy repo. I found a dependency issue affecting the other components.

eabdullin1 commented:
Please try this one
https://build.almalinux.org/build/42198

figless commented:
this is working well for me :)

eabdullin1 commented:
I published autopatch-core to the synergy reposittory. And also I updated it to the latest commit
